Mathlib/Data/Finset/PiAntidiagonal.lean

```diff
@@ -0,0 +1,392 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir and María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández, Bhavik Mehta
+-/
+
+import Mathlib.Data.Finset.Antidiagonal
+import Mathlib.Data.Finsupp.Antidiagonal
+import Mathlib.Data.Finsupp.Defs
+import Mathlib.Data.Finsupp.Interval
+
+/-! # Partial HasAntidiagonal for functions with finite support
+
+Let `μ` be an AddCommMonoid.
+
+In `Mathlib.Data.Finset.Antidiagonal` is defined a TypeClass
+`HasAntidiagonal μ` which provides a function `μ → Finset (μ × μ)
+which maps `n : μ` to a `Finset` of pairs `(a,b)`
+such that `a + b = n`.
+
+These functions apply to `ι →₀ ℕ`, more generally to `ι →₀ μ`
+under the additional assumption `OrderedSub μ` that make it
+a canonically ordered additive monoid.
+In fact, we just need an AddMonoid with a compatible order,
+finite Iic, such that if a + b = n, then a, b ≤ n,
+and any other bound would be OK.
+
+In this file, we provide an analogous definition for `ι →₀ μ`,
+with an explicit finiteness condition on the support.
+This Finset could be viewed inside `ι → μ`, but the `Finsupp` condition
+provides a natural `DecidableEq` instance.
+
+Consider types `ι` and `μ`, with `AddCommMonoid μ`.
+
+* The class `Finset.HasPiAntidiagonal ι μ` provides a finite set
+  `Finset.HasPiAntidiagonal.piAntidiagonal s n` of all functions
+  with finite support contained in `s` and sum `n : μ`
+  That condition is expressed by `HasPiAntidiagonal.mem_piAntidiagonal`
+* `Finset.HasPiAntidiagonal.mem_piAntidiagonal'` rewrites the `Finsupp.sum`
+  condition as a `Finset.sum`
+* Assuming `Finset.HasAntidiagonal μ`, we provide a member
+  `Finset.HasAntidiagonal.HasPiAntidiagonal` of that class
+* The construction starts with `Finset.HasPiAntidiagonal.finAntidiagonal`,
+  a variant of `Finset.Nat.antidiagonalTuple`
+
+-/
+
+namespace Finset
+
+open scoped BigOperators
+
+open Function Finsupp
+
+section HasPiAntidiagonal
+
+/-- The HasPiAntidiagonal class -/
+class HasPiAntidiagonal (ι : Type*) (μ : Type*) [AddCommMonoid μ] where
+  /-- The piAntidiagonal function -/
+  piAntidiagonal : Finset ι → μ → Finset (ι →₀ μ)
+  /-- A function belongs to `piAntidiagonal s n`
+    iff its support is contained in s and the sum of its components is equal to `n` -/
+  mem_piAntidiagonal {s} {n} {f} :
+    f ∈ piAntidiagonal s n ↔ f.support ≤ s ∧ sum f (fun _ x => x) = n
+
+export HasPiAntidiagonal (piAntidiagonal mem_piAntidiagonal)
+
+
+namespace HasPiAntidiagonal
+
+section Fin
+
+variable {μ : Type*} [AddCommMonoid μ] [DecidableEq μ] [HasAntidiagonal μ]
+
+/-- `finAntidiagonal d n` is the type of d-tuples with sum n -/
+noncomputable def finAntidiagonal : (d : ℕ) → μ → Finset (Fin d →₀ μ)
```

Comment:
I assume `Finsupp.cons` is what makes this `noncomputable`? I think for this function it would be best to stay computable and return
```suggestion
def finAntidiagonal : (d : ℕ) → μ → Finset (Fin d → μ)
```
Since it's easy to convert to finsupp later.

Reply:
Done. For the moment I've kept a Finsupp variant (with an index 0)